File: Teoricas/10.ml
(*Escopo Estatico
Nome da regra de resolução de nomes usada na maioria 
das linguagem modernas, incluindo o OCaml, C, C++ e Java.

Regra: Um uso dum nome refere-se sempre a ligacao sintaticamente envolvente, mais proxima.

No exemplo abaixo vemos a aplicacao dessa regra em Ocaml*)
let z = 5 in
    let f x = x + z in
        let z = 6 in
            f 0 (*o "z" a ser usado e o da linha 1, pq o uso do nome z 
                  refere-se a declaracao de z exterior (ou seja, envolvente).*)

let a = 5;;
let f x = x + a;;
let a = 10000;;
f 0

(*Escopo Dinamico
Nome da regra de resolução de nomes atualmente em desuso
Por ser mais confuso para o programador, enntrou em desuso

Regra: Um uso dum nome refere-se sempre a ligacao mais recentemente estabelecida
 para esse nome, temporalmente, durante a execucao do programa.
 
 No caso da funcao da linha 8, o resultado seria 6, pois era usado o valor mais
 recente de z (6)*)

File: Teoricas/10.c
#include <stdio.h>
/*Ligacoes (Bindings)
    Uma ligacao consiste numa associacao entre duas entidades.
    entidades podem ser nomes, localizacoes de memoria, tipos, objetos, etc.

    Ex: x=5;


Momento de estabelecimento duma ligação

    Em JS a ligacao e toda feita em tempo de execucao.

    -Quanto mais tarde se estabelecem as ligacoes mais flexivel e a linguagem.
    -Quanto mais cedo se estabelecem as ligacoes, mais rapida e a linguagem.

    1- a.f(x) vs 2 - f x
    O 1 apesar de ser mais lento, permite maior flexibilidade (Object orientend)


Classificao das ligacoes

    -Estaticas: efetuadas antes da execução do programa
    -Semidinamicas: efetuadas em tempo de execução mas determinadas em grande parte antes de o programa começar a correr.
        Em java os nomes das funcoes sao substituidos pela posicao da funcao na tabela com as varias funcoes do objeto.
    -Dinamicas: efetuadas completamente em tempo de execução.


Tempo de vida de uma ligacao

    Periodo de tempo da execucao dum programa durante a qual essa ligacao persiste.
    As ligações estaticas persistem durante a execução de TODO o programa. 
    As ligações semidinamicas e dinamicas persistem geralmente apenas durante PARTE da execucao do programa.
*/

/*Ambiente (conjunto de ligacoes para nomes)
    Todos os nomes que sao visiveis num determinado ponto do programa

Ambito (escopo) duma ligacao
    Regiao do programa na qual esse nome tem os atributos estabelecidos pela declaracao que introduz a ligacao.
    Os blocos "{ }" sao ambitos

Resolucao de nomes
    processo de descoberta do significado (ou seja, da ligacao) de 
    alguns nomes num ponto do programa onde esses nomes sao usados.
*/

int a; //variavel global, visivel em todo o programa
int main(void){
    int i; //i e acessivel ate a chaveta, mas tem um "buraco" pois e redefinido dentro das chavetas interiores
    int b = 5; //b so e acessivel apartir deste ponto
    i = a + b;
    {
        int i = 0;
        int j = i + 2; //o compilador tem que saber qual o "i" que esta a ser usado.
                        // faz isso atravez da resolucao de nomes
        int k = i + j;
        printf("%d %d %d\n", i, j, k);
    }
    printf("%d %d\n", i, b); 	
    return 0;
}